### PR TITLE
Some Sampling Ops

### PR DESCRIPTION
**Pull request by wchan**
issue: https://github.com/tensorflow/tensorflow/issues/565

This git pull request will give support to doing the sampling trick presented by Bengio et al., 2015 "Scheduled Sampling for Sequence Prediction with Recurrent Neural Networks". Originally I was thinking of doing the entire "softmax sampling" in one step (i.e., 1 kernel), but I decided to seperate it. I might submit a "monolithic" op for that later, but i highly doubt theres any major perf diff...

### CODE REVIEW

File: tensorflow/python/ops/sampling_ops.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import sys
+import numpy as np
+
+from tensorflow.python.framework import ops
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.framework import tensor_util
+# pylint: disable=wildcard-import
+# 'Constant' gets imported in the module 'array_ops'.
+from tensorflow.python.ops import gen_sampling_ops
```
Comment (ebrevdo) [inline tensorflow/python/ops/sampling_ops.py]: also:

`from tensorflow.python.ops.gen_sampling_ops import *`

any non-hidden ops should be imported directly as well.

File: tensorflow/python/kernel_tests/sampling_ops_test.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.python.ops import sampling_ops
+from tensorflow.python.ops import gen_sampling_ops
```
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/sampling_ops_test.py]: see comment below; here you should be able to call sampling_ops.bernoulli_sample, etc.

**Comment by vrv** [comment]
status of this PR?  you also need to add license headers to all of these new files

File: tensorflow/core/ops/sampling_ops.cc [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+REGISTER_OP("BernoulliSample")
+    .Input("p: float")
+    .Input("a: T")
+    .Input("b: T")
+    .Output("output: T")
+    .Attr("T: numbertype")
+    .Attr("seed: int = 0")
+    .Attr("seed2: int = 0")
+    .Doc(R"doc(
+Sample `a` with probability `p, or equivalently sample `b` with probability (1 - `p`).
```
Comment (gouwsmeister) [inline tensorflow/core/ops/sampling_ops.cc]: Missing closing ` after p.

File: tensorflow/core/ops/sampling_ops.cc [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+Sample `a` with probability `p`, or equivalently sample `b` with probability
+(1 - `p`).
+
+p: The probability to sample `a`.
+a: Heads.
+b: Tails.
+output: The sampled results which is either `a` or `b`.
+)doc");
+
+REGISTER_OP("SampleDistributionIndex")
```
Comment (ebrevdo) [inline tensorflow/core/ops/sampling_ops.cc]: Is this ever better than using the Box-Mueller transform?

Comment (ebrevdo) [inline tensorflow/core/ops/sampling_ops.cc]: Ah nevermind; I guess that only works for continuous, exponential family, distributions.

**Comment by gouwsmeister** [comment]
This looks good overall. Remember to add the required license headers as vrv pointed out.

**Comment by wchan** [comment]
whats the required license header? should i include the copyright google?
or just the apache part?

--
William Chan
Carnegie Mellon University
(650) 450-9455
williamchan.ca

**Comment by vrv** [comment]
Look at just about any other checked-in file :)

**Comment by wchan** [comment]
can someone pls launch a Jenkins, test this please.

--
William Chan
Carnegie Mellon University
(650) 450-9455
williamchan.ca

**Comment by vrv** [comment]
@tensorflow-jenkins test this please !!

File: tensorflow/python/ops/sampling_ops.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+# pylint: disable=wildcard-import
+from tensorflow.python.ops.gen_sampling_ops import *
+
+
+@ops.RegisterShape("BernoulliSample")
+def _BernoulliSampleShape(op):
+  a_shape = op.inputs[1].get_shape().with_rank(1)
+  b_shape = op.inputs[2].get_shape().with_rank(1)
+
+  assert a_shape == b_shape
```
Comment (ebrevdo) [inline tensorflow/python/ops/sampling_ops.py]: don't use simple "assert" statements ever; raise a ValueError or use the a_shape.assert_* methods

File: tensorflow/python/ops/sampling_ops.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+import sys
+import numpy as np
+
+from tensorflow.python.framework import ops
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.framework import tensor_util
+# pylint: disable=wildcard-import
+from tensorflow.python.ops.gen_sampling_ops import *
+
```
Comment (ebrevdo) [inline tensorflow/python/ops/sampling_ops.py]: pylint: enable=wildcard-import

**Comment by ebrevdo** [comment]
may want to add sample_ops to gen_docs ?

**Comment by vrv** [comment]
is this ready to review again?

**Comment by wchan** [comment]
I think we can remove this PR / ops... made aware of random_ops.multinomial which sorta make these ops obsolete.

**Comment by vrv** [comment]
Ah, gotcha.  We can reopen later if needed.  Thanks!
